src/turret.cpp

```diff
@@ -0,0 +1,382 @@
+#include "vehicle.h"
+
+#include "game.h"
+#include "player.h"
+#include "item.h"
+#include "itype.h"
+#include "veh_type.h"
+#include "npc.h"
+#include "messages.h"
+#include "translations.h"
+#include "ui.h"
+
+static const itype_id fuel_type_battery( "battery" );
+const efftype_id effect_on_roof( "on_roof" );
+
+std::vector<vehicle_part *> vehicle::turrets()
+{
+    std::vector<vehicle_part *> res;
+
+    for( auto &e : parts ) {
+        if( e.hp > 0 && e.base.is_gun() ) {
+            res.push_back( &e );
+        }
+    }
+    return res;
+}
+
+std::vector<vehicle_part *> vehicle::turrets( const tripoint &target )
+{
+    std::vector<vehicle_part *> res = turrets();
+
+    // exclude turrets not ready to fire or where target is out of range
+    res.erase( std::remove_if( res.begin(), res.end(), [&]( const vehicle_part * e ) {
+        return turret_query( *e ) != turret_status::ready ||
+               rl_dist( global_part_pos3( *e ), target ) > e->base.gun_range();
+    } ), res.end() );
+
+    return res;
+}
+
+vehicle::turret_status vehicle::turret_query( const vehicle_part &pt ) const
+{
+    if( !pt.base.ammo_sufficient() ) {
+        return turret_status::no_ammo;
+    }
+
+    auto ups = pt.base.get_gun_ups_drain() * pt.base.gun_current_mode().qty;
+    if( ups > fuel_left( fuel_type_battery ) ) {
+        return turret_status::no_power;
+    }
+
+    return turret_status::ready;
+}
+
+int vehicle::turret_fire( vehicle_part &pt )
+{
+    int shots = 0;
+
+    item &gun = pt.base;
+    if( !gun.is_gun() ) {
+        return false;
+    }
+
+    turret_reload( pt );
+
+    switch( turret_query( pt ) ) {
+        case turret_status::no_ammo:
+            add_msg( m_bad, string_format( _( "The %s is out of ammo" ), pt.name().c_str() ).c_str() );
+            break;
+
+        case turret_status::no_power:
+            add_msg( m_bad, string_format( _( "The %s is not powered" ), pt.name().c_str() ).c_str() );
```

Comment:
Also, why the weird formatting? Why `string_format` inside `add_msg` which already does formatting?